### PR TITLE
errhan: implement session errhan functions

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Implement  
 * MPI_Session_create_errhandler,
 * MPI_Session_set_errhandler
 * MPI_Session_get_errhandler
 * MPI_Session_call_errhandler

Minor cleanup of python binding scripts.

Some refactoring and code cleanup.

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/mpi/init/mpir_init.c [hunk 1]
```diff
@@ -111,6 +111,16 @@ int MPII_Init_thread(int *argc, char ***argv, int user_required, int *provided,
...
         MPIR_ERR_CHKHANDLEMEM(*p_session_ptr);
+
+        (*p_session_ptr)->errhandler = NULL;
+        /* FIXME: actually do something with session thread_level */
+        (*p_session_ptr)->thread_level = user_required;
+
+        {
+            int thr_err;
+            MPID_Thread_mutex_create(&(*p_session_ptr)->mutex, &thr_err);
+            MPIR_Assert(thr_err == 0);
```
Comment (raffenet) [inline src/mpi/init/mpir_init.c:122]: where is this mutex destroyed? it's not actually used in this commit, either.

File: src/binding/fortran/use_mpi_f08/mpi_f08_callbacks.f90 [hunk 2]
```diff
@@ -94,23 +94,27 @@ subroutine MPI_Type_delete_attr_function(datatype,type_keyval, &
 
 subroutine MPI_Comm_errhandler_function(comm,error_code)
     use mpi_f08_types, only : MPI_Comm
-    use mpi_f08_compile_constants, only : MPI_ADDRESS_KIND
```
Comment (raffenet) [inline src/binding/fortran/use_mpi_f08/mpi_f08_callbacks.f90:97]: removing these are unrelated

File: maint/local_python/binding_common.py [hunk 3]
```diff
@@ -72,14 +72,14 @@ def get_C_param(param, mapping):
...
-        param_type = re.sub(r'MPI_', 'MPIX_', param_type)
-
     if param['func_type']:
         param_type = param['func_type']
         if mapping['_name'].startswith("BIG_"):
             param_type += "_c"
 
+    if param_type in G.mpix_symbols:
+        param_type = re.sub(r'MPI_', 'MPIX_', param_type)
+
```
Comment (raffenet) [inline maint/local_python/binding_common.py:82]: Yeah, I suspected this wasn't actually working. I wrote some tests for partitioned that worked with both MPI and MPIX prefixes, but I forgot to report it...oops.

**Review by raffenet — COMMENTED** [review]
Looks mostly good. A couple of nitpicks.

File: src/mpi/init/mpir_init.c (hunk 1 above)
Comment (raffenet) [reply src/mpi/init/mpir_init.c:122]: I see the destroy code was added in a later commit. Not using it in the commit it is added is fine, to be clear.

File: src/binding/fortran/use_mpi_f08/mpi_f08_callbacks.f90 (hunk 2 above)
Comment (hzhou) [reply src/binding/fortran/use_mpi_f08/mpi_f08_callbacks.f90:97]: Split into a separate commit.

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: intel

File: src/mpi/init/mpir_init.c (hunk 1 above)
Comment (hzhou) [reply src/mpi/init/mpir_init.c:122]: Moved the destroy of mutex to the same commit.

**Review by raffenet — APPROVED** [review]

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
compiler: intel
